# Task 4.7

# Import Libraries

In [2]:
# Import Libraries

import pandas as pd
import numpy as np
import os

In [4]:
# setting path
path = r'/Users/shikhakamboj/Documents/JAN-2025 Instacart Basket Analysis'

# import data sets
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged.pkl'))

In [6]:
# subset created for instacart data
df = ords_prods_merge[:1000000]

In [8]:
df.shape

(1000000, 15)

In [57]:
# define a function for price labels

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range product'
  else: return 'Not enough data'

In [63]:
# apply the function

df['price_range'] = df.apply(price_label, axis=1)

In [61]:
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product     674229
Low-range product     312859
High range product     12912
Name: count, dtype: int64

In [16]:
df['prices'].max()

np.float64(99999.0)

In [20]:
df = df.copy()

In [22]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [24]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [26]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [40]:
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     674229
Low-range product     312859
High-range product     12912
Name: count, dtype: int64

In [30]:
# Let's repeat above steps using ords_prods_merge

ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [32]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [34]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [38]:
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [42]:
# printing the frequency of the “orders_day_of_week” column
ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [44]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [46]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly 

In [48]:
# Created new column in ords_prods_merge to view results

ords_prods_merge['busiest day'] = result

In [50]:
ords_prods_merge['busiest day'].value_counts(dropna= False)

busiest day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [52]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

# TASK 4.7

In [97]:
# Step 2  Created two busiest days of the week and two slowest days

result_updated = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_updated.append("Busiest days")
  elif value == 3 or value == 4:
    result_updated.append("Slowest days")
  else:
    result_updated.append("Regularly busy")

In [74]:
result_updated

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',


In [77]:
# step 3 Created new column in ords_prods_merge to view results

ords_prods_merge['busiest days'] = result_updated

In [79]:
ords_prods_merge['busiest days'].value_counts(dropna= False)

busiest days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [99]:
# Check the values of this new column

ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0         2   202279             3                   5                  9   
1         2   202279             3                   5                  9   
2         2   202279             3                   5                  9   
3         2   202279             3                   5                  9   
4         2   202279             3                   5                  9   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     8.0     33120.0                1.0        1.0   
1                     8.0     28985.0                2.0        1.0   
2                     8.0      9327.0                3.0        0.0   
3                     8.0     45918.0                4.0        1.0   
4                     8.0     30035.0                5.0        0.0   

            product_name  aisle_id  department_id  prices match  \
0     Organic Egg Whites        86             16    11.3  both   
1  Michigan Organic Kale        83              4    13.4  both   
2          Garlic Powder       104             13     3.6  both   
3         Coconut Butter        19             13     8.4  both   
4      Natural Sweetener        17             13    13.7  both   

     price_range_loc     busiest day    busiest days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy  Regularly busy  
2  Low-range product  Regularly busy  Regularly busy  
3  Mid-range product  Regularly busy  Regularly busy  
4  Mid-range product  Regularly busy  Regularly busy

The total count in ords_prods_merge matches the sum of all groupings, with each label's total aligning with its qualifying days. For example, "Busiest days" (11,864,412) matches orders on days 0 and 1 (6,204,182 + 5,660,230).

In [95]:
# step 4 Create a new column containing these labels called “busiest_period_of_day.”

ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

The order_hour_of_day counts are split into: "Most orders" (10, 11, 14, 15, 13, 12, 16, 9), "Fewest orders" (23, 6, 0, 1, 5, 2, 4, 3), and "Average orders" (all remaining hours).

In [101]:
result_for_hours_of_day = []

for value in ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_for_hours_of_day.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_for_hours_of_day.append("Fewest orders")
  else:
    result_for_hours_of_day.append("Average orders")

In [103]:
result_for_hours_of_day

['Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',


In [105]:
#  Created new column in ords_prods_merge to view results

ords_prods_merge['busiest_period_of_day'] = result_for_hours_of_day

In [107]:
# Step 5 printing the frequency 

ords_prods_merge['busiest_period_of_day'].value_counts(dropna= False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [111]:
# step 7 Export dataframe as a pickle file

ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))